examples/cross-contract-call-tracing/lib.rs

```diff
@@ -0,0 +1,173 @@
+
+#[ink::contract]
+mod contract {
+    use ink::env::{
+        call::{build_call, ExecutionInput, Selector},
+        DefaultEnvironment,
+    };
+
+    #[ink(storage)]
+    pub struct CrossCallingContract;
+
+    impl CrossCallingContract {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {}
+        }
+
+        #[ink(message)]
+        pub fn outer_call(
+            &self,
+            next_callee: AccountId,
+            next_next_callee: AccountId,
+            arg: u32,
+        ) -> u32 {
+            build_call::<DefaultEnvironment>()
+                .call(next_callee)
+                .gas_limit(0)
+                .transferred_value(0)
+                .exec_input(
+                    ExecutionInput::new(Selector::new(ink::selector_bytes!("middle_call")))
+                        .push_arg(next_next_callee)
+                        .push_arg(arg),
+                )
+                .returns::<u32>()
+                .invoke()
+        }
+
+        #[ink(message)]
+        pub fn middle_call(&self, next_callee: AccountId, arg: u32) -> u32 {
+            build_call::<DefaultEnvironment>()
+                .call(next_callee)
+                .gas_limit(0)
+                .transferred_value(0)
+                .exec_input(
+                    ExecutionInput::new(Selector::new(ink::selector_bytes!("inner_call")))
+                        .push_arg(arg),
+                )
+                .returns::<u32>()
+                .invoke()
+        }
+
+        #[ink(message)]
+        pub fn inner_call(&self, arg: u32) -> u32 {
+            match arg % 2 {
+                0 => arg / 2,
+                _ => 3 * arg + 1,
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ink::storage::traits::Storable;
+    use std::{cell::RefCell, error::Error, fs, path::PathBuf, rc::Rc};
+
+    use drink::{
+        runtime::{
+            pallet_contracts_debugging::{DebugExt, DebugExtT},
+            MinimalRuntime,
+        },
+        session::{
+            contract_transcode::{ContractMessageTranscoder, Tuple, Value},
+            Session,
+        },
+        AccountId32,
+    };
+
+    fn transcoder() -> Rc<ContractMessageTranscoder> {
+        let path = PathBuf::from("target/ink/cross_contract_call_tracing.json");
+        Rc::new(ContractMessageTranscoder::load(path).expect("Failed to create transcoder"))
+    }
+
+    fn bytes() -> Vec<u8> {
+        let path = "target/ink/cross_contract_call_tracing.wasm";
+        fs::read(path).expect("Failed to find or read contract file")
+    }
+
+    fn ok(v: Value) -> Value {
+        Value::Tuple(Tuple::new(Some("Ok"), vec![v]))
+    }
+
+    thread_local! {
+        static OUTER_ADDRESS:  RefCell<Option<AccountId32>> = RefCell::new(None);
+        static MIDDLE_ADDRESS:  RefCell<Option<AccountId32>> = RefCell::new(None);
+        static INNER_ADDRESS:  RefCell<Option<AccountId32>> = RefCell::new(None);
```

Comment:
Can't we put these 3 inside of `TestDebugger`?

Reply:
yes, exactly; but we could put `TestDebugger` inside the testcase, but this makes things very opaque